components/cloud/gcp/src/kms.rs

```diff
@@ -0,0 +1,407 @@
+// Copyright 2024 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{fmt, result::Result as StdResult};
+
+use async_trait::async_trait;
+use cloud::{
+    error::{Error as CloudError, KmsError, Result},
+    kms::{Config, CryptographyType, DataKeyPair, EncryptedKey, KmsProvider, PlainKey},
+    metrics, KeyId,
+};
+use futures_util::stream::StreamExt;
+use http::Method;
+use hyper::Body;
+use regex::Regex;
+use serde::{Deserialize, Deserializer, Serialize};
+use tame_gcs::error::HttpStatusError;
+use tikv_util::{box_err, stream::RetryError, time::Instant};
+
+use crate::{
+    client::{GcpClient, RequestError},
+    STORAGE_VENDOR_NAME_GCP,
+};
+
+// generated random encryption data key length.
+const DEFAULT_DATAKEY_SIZE: usize = 32;
+// google kms endpoint.
+const GCP_KMS_ENDPOINT: &str = "https://cloudkms.googleapis.com/v1/";
+
+/// The encryption key_id pattern of gcp ksm:
+///   projects/{project_name}/locations/{location}/keyRings/{key_ring}/
+/// cryptoKeys/{key}
+const KEY_ID_PATTERN: &str =
+    r"^projects/([^/]+)/locations/([^/]+)/keyRings/([^/]+)/cryptoKeys/([^/]+)/?$";
+
+// following are related kms api method names:
+const METHOD_ENCRYPT: &str = "encrypt";
+const METHOD_DECRYPT: &str = "decrypt";
+const METHOD_GEN_RANDOM_BYTES: &str = "generateRandomBytes";
+
+/// Protection level of the generated random key, always using HSM(Hardware
+/// Security Module).
+const RANDOMIZE_PROTECTION_LEVEL: &str = "HSM";
+
+pub struct GcpKms {
+    config: Config,
+    // the location prefix of key id,
+    // format: projects/{project_name}/locations/{location}
+    location: String,
+    client: GcpClient,
+}
+
+impl GcpKms {
+    pub fn new(mut config: Config) -> Result<Self> {
+        assert!(config.gcp.is_some());
+        let re = Regex::new(KEY_ID_PATTERN).unwrap();
```

Comment:
Maybe you can put the regex in to lazy_static or OnceCell.

Reply:
done